sycl/unittests/kernel-and-program/Cache.cpp

```diff
@@ -0,0 +1,485 @@
+
+  globalCtx.reset(new TestCtx{CtxImpl->getHandleRef()});
+
+  program Prg{Ctx};
+
+  Prg.compile_with_kernel_type<TestKernel>("-g");
+  Prg.link();
+  kernel Ker = Prg.get_kernel<TestKernel>();
+  detail::KernelProgramCache::KernelCacheT &Cache =
+      CtxImpl->getKernelProgramCache().acquireKernelsPerProgramCache().get();
+  EXPECT_EQ(Cache.size(), 0) << "Expect empty cache for kernels";
+}
+
+TEST(KernelAndProgramCache, KernelCacheNegativeLinkOpts) {
+  platform Plt{default_selector()};
+  if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+    std::clog << "This test is only supported on OpenCL devices\n";
+    return;
+  }
+
+  unittest::PiMock Mock(Plt);
+  Mock.redefine<detail::PiApiKind::piProgramCompile>(redefinedProgramCompile);
+  Mock.redefine<detail::PiApiKind::piProgramLink>(redefinedProgramLink);
+  Mock.redefine<detail::PiApiKind::piKernelCreate>(redefinedKernelCreate);
+  Mock.redefine<detail::PiApiKind::piKernelRetain>(redefinedKernelRetain);
+  Mock.redefine<detail::PiApiKind::piKernelRelease>(redefinedKernelRelease);
+  Mock.redefine<detail::PiApiKind::piKernelGetInfo>(redefinedKernelGetInfo);
+
+  context Ctx{Plt};
+  auto CtxImpl = detail::getSyclObjImpl(Ctx);
+
+  globalCtx.reset(new TestCtx{CtxImpl->getHandleRef()});
+
+  program Prg{Ctx};
+
+  Prg.compile_with_kernel_type<TestKernel>();
+  Prg.link("-g");
+  kernel Ker = Prg.get_kernel<TestKernel>();
+  detail::KernelProgramCache::KernelCacheT &Cache =
+      CtxImpl->getKernelProgramCache().acquireKernelsPerProgramCache().get();
+  EXPECT_EQ(Cache.size(), 0) << "Expect empty cache for kernels";
+}
+
+TEST(KernelAndProgramCache, KernelCacheNegativeLinkedProgs) {
+  platform Plt{default_selector()};
+  if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+    std::clog << "This test is only supported on OpenCL devices\n";
+    return;
+  }
+
+  unittest::PiMock Mock(Plt);
+  Mock.redefine<detail::PiApiKind::piProgramCompile>(redefinedProgramCompile);
+  Mock.redefine<detail::PiApiKind::piProgramLink>(redefinedProgramLink);
+  Mock.redefine<detail::PiApiKind::piKernelCreate>(redefinedKernelCreate);
+  Mock.redefine<detail::PiApiKind::piKernelRetain>(redefinedKernelRetain);
+  Mock.redefine<detail::PiApiKind::piKernelRelease>(redefinedKernelRelease);
+  Mock.redefine<detail::PiApiKind::piKernelGetInfo>(redefinedKernelGetInfo);
+
+  context Ctx{Plt};
+  auto CtxImpl = detail::getSyclObjImpl(Ctx);
+
+  globalCtx.reset(new TestCtx{CtxImpl->getHandleRef()});
+
+  program Prg1{Ctx};
+  program Prg2{Ctx};
+
+  Prg1.compile_with_kernel_type<TestKernel>();
+  Prg2.compile_with_kernel_type<TestKernel2>();
+  program Prg({Prg1, Prg2});
+  kernel Ker = Prg.get_kernel<TestKernel>();
+
+  detail::KernelProgramCache::KernelCacheT &Cache =
+      CtxImpl->getKernelProgramCache().acquireKernelsPerProgramCache().get();
+  EXPECT_EQ(Cache.size(), 0) << "Expect empty cache for kernels";
+}
+
+TEST(KernelAndProgramCache, KernelCacheNegativeSource) {
```

Comment:
Just curious: I see that there are fourteen tests while the previous version contains only ten tests. What test cases have been added?

Reply:
Source-based programs with build/link/compile options.